Add workers settings to configuration manual (#14086)
matrix-org/synapse: docs/usage/configuration/config_documentation.md

        false
```
---

> ## Worker configuration for main process

Reviewer: ```suggestion
## Coordinating workers
```

The worker configuration is divided into two parts.

1. These switches for the shared configuration to give the main process
   the necessary information.
1. And the [switches for the worker configuration file](#worker-configuration-for-workers-configuration)
   to configure each worker them self are futher below.

The manual how to configure workers is described in detail in the
[worker documentation](../../workers.md)

---
### `worker_replication_secret`

A shared secret used by the replication APIs to authenticate HTTP requests
from workers.

By default this is unused and traffic is not authenticated.

Example configuration:
```yaml
worker_replication_secret: "secret_secret"
```
---
### `start_pushers`

Controls sending of push notifications on the main process. Set to `false`
if using a [pusher worker](../../workers.md#synapseapppusher). Defaults to `true`.

Example configuration:
```yaml
start_pushers: false
```
---
### `pusher_instances`

It is possible to run multiple [pusher workers](../../workers.md#synapseapppusher),
in which case the work is balanced across them. Use this setting to list the pushers by

> [`worker_name`](#worker_name).

Author: I don't know if that's the case.

> [`worker_name`](#worker_name).

Reviewer 2: That is correct, restart all pusher workers. For now. In the future, a SIGHUP will be enough to reset it's settings but that's way down the line.

> [`worker_name`](#worker_name).

Author: Is that the case? I would imagine that the main process needs to be restarted, as it distributes the pushes to the pusher_workers. When the worker receives the command, it initialises the push. For the single push worker, it is irrelevant whether other push requests were done by another worker.

> [`worker_name`](#worker_name).

Reviewer 2: You are most likely correct. I misread. I'm not the expert on this particular topic, I just always restart.

> [`worker_name`](#worker_name).

Reviewer: >  I would imagine that the main process needs to be restarted, as it distributes the pushes to the pusher_workers.

> For the single push worker, it is irrelevant whether other push requests were done by another worker.

I think these aren't quite correct. Looking at the source code, I think:

- _Every_ worker realises they need to send out a push by a) inspecting the redis replication streams or b) responding to in-process changes via the Notifier.
- Those which are pusher workers have `PusherPool._should_start_pushers` set to True, by inspecting the `_pusher_shard_config`, which reads from `pusher_instances` in the yaml.
- Each pusher worker checks to see if it should handle a push with `self._pusher_shard_config.should_handle(self._instance_name, user_id)`
- the answer depends on the hash of the user_id, the number of pushers, and the order of the pushers in `pusher_instances`
- adding a new pusher_instance will change how those answers are [calculated](https://github.com/matrix-org/synapse/blob/5e5c8150d798f6929ddedbb39f9f11486558cdbc/synapse/config/_base.py#L945-L954)
- which suggests that you need to restart **all** pusher workers whenever you change the `push_instances`.

(cc @matrix-org/synapse-core, please check my working)

> [`worker_name`](#worker_name).

Author: => https://github.com/matrix-org/synapse/pull/14086/commits/5045ca0edf89f691ca009fbac3d44194ddb22fe6

> If only one pusher worker is configured, this setting is not necessary.

Reviewer: Can we also say something like this?

```
If no pusher workers are configured, this setting is not necessary---but
[`start_pushers`](#start_pushers) should be set to `true` so that the
main process will send out push notifications.
```

> If only one pusher worker is configured, this setting is not necessary.

Reviewer 2: [This](https://github.com/matrix-org/synapse/blob/develop/synapse/config/workers.py#L172) logic says that method is outdated. See [efforts](https://github.com/matrix-org/synapse/pull/14294) to change this.

> If only one pusher worker is configured, this setting is not necessary.

Author: The comment (for the `federation_sender`) in the code is interesting. I try to document the current status here. If the logic has changed, it needs a new PR to update the documentation, deprecate the old federation_sender and remove the old code.

> If only one pusher worker is configured, this setting is not necessary.

Reviewer: > I think, something like this is better

Sure, but let's tweak the last sentence to say: "if you do not disable it by setting ...".

> If only one pusher worker is configured, this setting is not necessary.

Author: => https://github.com/matrix-org/synapse/pull/14086/commits/5045ca0edf89f691ca009fbac3d44194ddb22fe6

Example configuration:
```yaml
pusher_instances:
  - pusher_worker1

>   - pusher_worker2

Author: Good idea. I have added it.

>   - pusher_worker2

Reviewer 2: It's unnecessary. See comment above.

>   - pusher_worker2

Reviewer: I don't have strong opinions other than: I would rather we deprecate `start_pushers` rather than deal with having two related options interact in a confusing way. But that's not something to deal with here.

```
---
### `send_federation`

Controls sending of outbound federation transactions on the main process.

> Set to false if using a [federation sender worker](../../workers.md#synapseappfederation_sender).

Reviewer: ```suggestion
Set to `false` if using a [federation sender worker](../../workers.md#synapseappfederation_sender).
```

Reviewer 2: Same as `pushers`. The logic is the same. If it's on it's `federation_sender_instances` map, you don't this at all. [Here](https://github.com/matrix-org/synapse/blob/develop/synapse/config/workers.py#L172)

        false
```
---
## Coordinating workers

> Configuration options related to workers.

Reviewer: Would this help clear up the difference between the two sections?

```suggestion
Configuration options related to workers which belong in the main config file
(usually called `homeserver.yaml`).
```

---
### `federation_sender_instances`

It is possible to run multiple
[federation sender worker](../../workers.md#synapseappfederation_sender), in which
case the work is balanced across them. Use this setting to list the senders.

This configuration setting must be shared between all federation sender workers, and if

> changed all federation sender workers must be stopped at the same time and then

Reviewer: As above: should the example below also include `send_federation: false`?

Author: Good idea. I have added it.

Example configuration:
```yaml
send_federation: false
federation_sender_instances:
  - federation_sender1
```
---
### `instance_map`

> When using workers this should be a map from [`worker_name`](#worker_name) to the
> HTTP replication listener of the worker, if configured.
> Only `synapse.app.generic_worker` as [`stream_writers`](../../workers.md#stream-writers)
> needs a HTTP replications listener. It must also be specified the
> [`instance_map`](#instance_map).

Reviewer: Doesn't the main process also need an HTTP replication listener?

```suggestion
When using workers this should be a map from [`worker_name`](#worker_name) to the
HTTP replication listener of the worker, if configured.
Each worker declared under [`stream_writers`](../../workers.md#stream-writers) needs a HTTP replication listener, and that listener should be included in the `instance_map`. (The main process also needs an HTTP replication listener, but that should not be listed in the `instance_map`.)
```

---
### `instance_map`

When using workers this should be a map from [`worker_name`](#worker_name) to the
HTTP replication listener of the worker, if configured.

> Not every type of worker needs a HTTP replications listener.

Reviewer: Can we say _which_ workers need a replication listener here? Is this any worker which accepts requests from other workers? E.g. event persister, pushers? Maybe we should refer to `stream_writers`?

Is this (should this be) explained in the worker docs page?

Author: See: https://github.com/matrix-org/synapse/pull/14086/commits/17ec5d3cf1a6398480647753ed466f955d2466ee

Experimental: When using workers you can define which workers should

> handle event persistence and typing notifications. Any worker

Reviewer: I think this is out of date, looks like we have more streams:

https://github.com/matrix-org/synapse/blob/32fc3b7ba4702a0068a82bdd0595e2f426967d4d/synapse/config/workers.py#L72-L113

Reviewer 2: It is out-of-date. See 
* #14197 

for the PR with additional information.

Author: I have reword this: https://github.com/matrix-org/synapse/pull/14086/commits/a3ad96aefc518e3faf85de406df36a92f83d7c50
The list of stream writers is currently part of the [worker documentation](https://matrix-org.github.io/synapse/latest/workers.html#stream-writers).

Reviewer: Thanks, I think this is great.

### `stream_writers`

Experimental: When using workers you can define which workers should

> handle write streams such as event persistence and typing notifications.

Reviewer: Perhaps "handling writing to streams" here>

Author: => https://github.com/matrix-org/synapse/pull/14086/commits/a214ee89e8c0702c585897484a58fd5b9c94f0fd

Reviewer: Oops, typo from me. I should have said "handle writing to streams". I'll fixup

Reviewer: 8394c3eb7d

  port: 6379
  password: <secret_password>
```
---

> ## Worker configuration for workers configuration

Author: I am happy to receive better suggestions

> These switches configure each worker them self with the a worker configuration file.

Reviewer: ```suggestion
These options configure an individual worker, in its worker configuration file.
```

Perhaps even:
```suggestion
These options configure an individual worker, in its worker configuration file.
They should be not be provided when configuring the main process.
````

> Note also the configuration for the
> [main process above](#worker-configuration-for-main-process).

Reviewer: As I suggested elsewhere: I don't think the previous batch of settings configure just communication with the main worker? I've suggested language like: they configure communication between "ordinary" and "privileged" workers.

Perhaps:

```suggestion
Note also the configuration above for
[coordinating a cluster of workers](#coordinating-workers).
```

> The manual how to configure workers is described in detail in the
> [worker documentation](../../workers.md)

Reviewer: ```suggestion
For guidance on setting up workers, see the [worker documentation](../../workers.md).
```

as suggested above

---

> ### `worker_app`

Reviewer: Is this required? Does it have a default value?

Author: IMO it is required and there is no default. But I am not sure.
Ther is a related issue: https://github.com/matrix-org/synapse/issues/13644#issue-1352303831

Reviewer 2: This is correct. No default.

Reviewer: Looking at the source it looks like Synapse assumes it is the main process if this value is not provided (or if it is provided and set to `synapse.app.homeserver`).

Wouldn't object to us saying this is required, but I think the text is also fine as it is.

The type of worker. The currently available worker applications are listed
in [worker documentation](../../workers.md#available-worker-applications).

The most common worker is the
[`synapse.app.generic_worker`](../../workers.md#synapseappgeneric_worker).

Example configuration:
```yaml
worker_app: synapse.app.generic_worker
```
---

> ### `worker_name`

Reviewer: Presumably this is required---no default?

We should probably say that worker names need to be unique---things will break if two workers have the same name!

Author: There is a default: https://github.com/matrix-org/synapse/blob/0d59ae706a09e95c500a594159fff992260147ac/synapse/config/workers.py#L155
IMO it is only needed if you delegate some tasks to the worker, e.g. `stream_writer`. For a simple `generic_worker` it is not needed.
Related issue: https://github.com/matrix-org/synapse/issues/8084

Reviewer: > IMO it is only needed if you delegate some tasks to the worker, e.g. stream_writer.

This seems plausible.

We could maybe say "we strongly recommend giving each worker a unique `worker_name`", but I'm also happy to leave this as it is.

Author: => https://github.com/matrix-org/synapse/pull/14086/commits/c0434bde3b8acbbb283e243fd78e35f85177b1e0

A unique name for the worker. The worker needs a name to be addressed in
further parameters and identification in log files.

Example configuration:
```yaml
worker_name: generic_worker1
```
---
### `worker_replication_host`

The HTTP replication endpoint that it should talk to on the main Synapse process.
The main Synapse process defines this with a `replication` resource in
[`listeners` option](#listeners).

Example configuration:
```yaml
worker_replication_host: 127.0.0.1
```
---
### `worker_replication_http_port`

The HTTP replication port that it should talk to on the main Synapse process.
The main Synapse process defines this with a `replication` resource in
[`listeners` option](#listeners).

Example configuration:
```yaml
worker_replication_http_port: 9093

> ```

Reviewer: Do these settings have defaults?

Author: No default
https://github.com/matrix-org/synapse/blob/32fc3b7ba4702a0068a82bdd0595e2f426967d4d/synapse/config/workers.py#L150

Reviewer: Thanks for checking.

---
### `worker_listeners`

> A worker can handle HTTP requests. If handling HTTP requests, a `worker_listeners`

Reviewer: I think this should read: "Some workers can handle internal HTTP requests from other workers."? Maybe even "Some privileged workers"?

Author: The workers can not handle internal request from other workers. They can handle external request from clients or other federations.

Reviewer: > The workers can not handle internal request from other workers.

I don't think this is correct: e.g. event_creators make HTTP calls to event_persisters using `/_synapse/replication/send_event/EVENT_ID/TXN`, see [here](https://github.com/matrix-org/synapse/blob/8ab16a92edd675453c78cfd9974081e374b0f998/synapse/replication/http/send_event.py#L116-L154). In general I think `worker_listeners` needs to include a replication listener here if you're one of the workers in the `instance_map`/`stream_writers`.

But: you're absolutely right that you need to define e.g. client/federation listeners here too. I don't know if we ought to mention the replication listeners here or if that will cause further confusion.

How about:

```
A worker can handle HTTP requests. To do so, a `worker_listeners` option 
must be declared, in the same way as the [`listeners` option](#listeners) 
in the shared config.

Workers declared in [`stream_writers`](#LINK_TODO) will need to include a
`replication` listener here, in order to accept internal HTTP requests from
other workers.
```

Author: => https://github.com/matrix-org/synapse/pull/14086/commits/a943440797467342ad132cadeb5d98647fc4ab53

option with an http listener, in the same way as the [`listeners` option](#listeners)
in the shared config.

Example configuration:
```yaml
worker_listeners:
  - type: http
    port: 8083
    resources:
      - names: [client, federation]
```
---
### `worker_daemonize`

> Specifies whether the worker should be daemonize. If
> [systemd](../../systemd-with-workers/README.md) is used, this must not configured.
> Systemd manages daemonization itself. Defaults to `false`.

Reviewer: ```suggestion
Specifies whether the worker should be started as a [daemon process](TODO_FIND_A_LINK).
If Synapse is being managed by [systemd](../../systemd-with-workers/README.md), this option 
must be omitted or set to `false`.

Defaults to `false`.
```

Example configuration:
```yaml
worker_daemonize: true
```

> ---

Reviewer: Can you briefly explain what daemonization is (e.g. a link to wikipedia or some similar reference)?

Author: Honestly, no. And I didn't find anything about it in the Synapse docs.

Reviewer 2: The short answer is if you have a process orchestrating(like `systemd` or `supervisord` the start and stop of the individual workers and master process, then you don't want to use this. If you are running, for example, on multiple machines and start the process with scripting starting each process or `synctl` then this is useful. The [Wikipedia](https://en.wikipedia.org/wiki/Daemon_(computing)) entry is exhaustive reading and not particularly useful when someone just wants to know "Do I need this?"

Reviewer: Thanks both. Let's leave this as it is.

### `worker_pid_file`

> When running Synapse worker as a daemon, the file to store the pid in. Defaults to none.
> This is the same way as the[`pid_file` option](#pid_file) in the shared config.

Reviewer: ```suggestion
When running a worker as a daemon, we need a place to store the 
[PID](https://en.wikipedia.org/wiki/Process_identifier) of the worker.
This option defines the location of that "pid file".

This option is required if `worker_daemonize` is `true` and ignored 
otherwise. It has no default.

See also the [`pid_file` option](#pid_file) option for the main Synapse process.
```

Example configuration:
```yaml

> worker_pid_file: DATADIR/generic_worker1.pid

Reviewer: is DATADIR a special path interpreted by Synapse here?

Author: No. This is a placeholder that the docs also use elsewhere. This seems to be a linux default.

Reviewer: Seems to come from https://github.com/matrix-org/synapse/blob/6faaf76a3239c3fd1a9adc0b45b04fcf43237824/synapse/_scripts/generate_config.py#L19-L24

Don't think this is a Linux thing per se.

```
---
### `worker_log_config`

This option specifies a yaml python logging config file as described
[here](https://docs.python.org/3.11/library/logging.config.html#configuration-dictionary-schema).

> This is the same way as the [`log_config` option](#log_config) in the shared config.

Reviewer: ```suggestion
See also the [`log_config` option](#log_config) option for the main Synapse process.
```

run_background_tasks_on: worker1
```
---

> ### `worker_replication_secret`

Author: I have moved this further up.
